sp_mock/backend/server.js

```diff
@@ -136,6 +138,14 @@ app.post("/api/login/layer8/auth", async (req, res) => {
   );
   res.status(200).json({ token });
 });
+
+app.post("/api/upload", async (req, res) => {
+  const file = req.body.get('file');
+  const uint8Array = new Uint8Array(await file.arrayBuffer());
+  fs.writeFileSync(`./uploads/${file.name}`, uint8Array);
+  res.status(200).json({ message: "File uploaded successfully!" });
+});
+
```

Comment:
This method will not work outsid Layer8. 

@stravid87 correct me if i'm wrong. 

With layer 8 we should be able to have the same feature with layer8 enable or not. 
So the final user don't have to write two function to handle image upload. 

Here your function `/api/upload` is getting the image in `req.body.get('file')`
But by defautl the image is inside `req.file` or `req.files`

Reply:
This completely depends on the backend. The data will come in as a `FormData` with a `multipart/formdata` content type, so you can read the data however you want. I did it this way just for you to see how it works

Reply:
If you check the type of `req.body`, you'll get a `FormData`. By referencing `file`, I'm just getting that field from the `FormData`

Reply:
> This method will not work outsid Layer8.
> 
> @stravid87 correct me if i'm wrong.
> 
> With layer 8 we should be able to have the same feature with layer8 enable or not. So the final user don't have to write two function to handle image upload.
> 
> Here your function `/api/upload` is getting the image in `req.body.get('file')` But by defautl the image is inside `req.file` or `req.files`

Oh Jesus: "The data sent by multipart/form-data is separated in parts divided by a boundary specified in the content-type header sent by the browser, an example below:"

```
------WebKitFormBoundaryxJi9AgGdxx83BunR
Content-Disposition: form-data; name="username""
```

I'm looking at this now and I must admit I'm at a loss. 

Looking at how "normally" nodejs handles multipart form data... it almost always seems to require a 3rd party package (i.e., multer) unless you want to get crazy (see: (https://plainenglish.io/blog/parsing-post-data-3-different-ways-in-node-js-e39d9d11ba8)).

@hermannleboss, when you say that "by default the image is inside req.file or req.files", are you using ExpressJS or are you referring to a Vanilla NodeJS implementation?

Reply:
Both. 


Here is an implementation of theses functions


```js

// app.js or app.ts
const express = require('express');
const multer = require('multer');
const fileUpload = require('express-fileupload');
const path = require('path');
const fs = require("fs");


const cors = require('cors')

const app = express();
const port = 3000;

app.use(cors())
app.use(fileUpload());

// Set up multer for handling file uploads
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, 'uploads/'); // Specify the directory where uploaded files will be stored
  },
  filename: (req, file, cb) => {
    cb(null, `${Date.now()}-${file.originalname}`); // Use a unique filename
  },
});

const upload = multer({ storage: storage });

// Serve the HTML form for testing
app.get('/', (req, res) => {
  res.sendFile(path.join(__dirname, 'index.html'));
});

// Handle file uploads
app.post('/upload', upload.single('file'), (req, res) => {
  // 'file' should match the name attribute in the HTML form
  
  console.log("Multer upload");
//   console.log("req.file", req.file);
//   console.log("req.files", req.files);
  console.log("req.body", req.body.formData);
  const uploadedFile = req.file;


  if (!uploadedFile) {
    return res.status(400).json({ error: 'No file uploaded' });
  }

  // Do something with the uploaded file, for example, send a success response
  res.json({ message: 'File uploaded successfully', filename: uploadedFile.filename });
});

// Handle file uploads
app.post('/upload/natif', (req, res) => {
    console.log("Native upload");
    console.log("req.file", req.file);
    console.log("req.files", req.files);
    console.log("req.body", req.body);
    try {
      if (!req.files || Object.keys(req.files).length === 0) {
        return res.status(400).json({ error: 'No file uploaded' });
      }
  
      const uploadedFile = req.files.file; // 'file' should match the name attribute in the HTML form
  
      // Specify the directory where uploaded files will be stored
      const uploadDir = path.join(__dirname, 'uploads');
      
      // Ensure the 'uploads' directory exists
      if (!fs.existsSync(uploadDir)) {
        fs.mkdirSync(uploadDir);
      }
  
      // Create a unique filename
      const filename = `${Date.now()}-${uploadedFile.name}`;
      const filePath = path.join(uploadDir, filename);
  
      // Save the file to the server
      uploadedFile.mv(filePath, (err) => {
        if (err) {
          return res.status(500).json({ error: 'Error saving the file' });
        }
  
        // Do something with the uploaded file, for example, send a success response
        res.json({ message: 'File uploaded successfully', filename });
      });
    } catch (err) {
      console.error(err);
      res.status(500).json({ error: 'Server error' });
    }
  });

  // Daniel method
  app.post("/api/upload", async (req, res) => {
    console.log(req.body);
    const file = req.body.get('file');
    const uint8Array = new Uint8Array(await file.arrayBuffer());
    fs.writeFileSync(`./uploads/${file.name}`, uint8Array);
    res.status(200).json({ message: "File uploaded successfully!" });
  });
app.listen(port, () => {
  console.log(`Server is running at http://localhost:${port}`);
});

```

In the native methode and multler methode the file are inside req.file or req.files object 

But with daniel methode it's not working.

Reply:
Daniels' wont work so long as there is a middleware in use. Specifically, note the use of `app.use(fileUpload());`
This call mutates the data and by the time it gets to the route `app.post("/api/upload")`, it isn't the same and the function will fail. 

Note that to make Multer work, you need to place `upload.single('file')` in the route handler. 
Also, note that to make express-fileupload work, you need to put the middleware. I assume Daniel could do something similar by hiding this logic in it's own function:
```
 const file = req.body.get('file');
 const uint8Array = new Uint8Array(await file.arrayBuffer());
 fs.writeFileSync(`./uploads/${file.name}`, uint8Array);
```

In general, this brings up a very important design consideration. I feel like, if we start designing custom solutions (that is replacing packages like Multer) there will be an overwhelming amount of work and it simply won't be possible to do. 

@DeeStarks, I know there is a way to make layer8 work transparently. Currently, my intuition says that by calling the following, we are very likely preventing our future use of extremely common middleware such as `app.use(express.json())`

``` [line 324, middleware.go]
    req.Call("on", "data", js.FuncOf(func(this js.Value, args []js.Value) interface{} {
		body += args[0].Call("toString").String()
		return nil
	}))
```

Reply:
Seeing @hermannleboss interact with layer8, @DeeStarks, I think we have a major design decision to make: do we can to integrate tightly with ExpressJS decreasing our flexibility but simplifying much? Or, do we want to make the middleware work with or _without_ ExpressJS ?

Reply:
@hermannleboss it would work. This is why I have transformed the request data into `multipart/formdata` before it gets to the handler

What I did on the backend is a simple implementation to show that it works. Try with multer or any other library that accepts formdata, it would work and you can do `req.file`

Reply:
Yes @stravid87, it would only work when you do `upload.single('file')` for now. We can factor in multiple files later